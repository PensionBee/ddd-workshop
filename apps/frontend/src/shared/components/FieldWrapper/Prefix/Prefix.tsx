import { SmallSystemIcon } from "@shared/components/Svg";
import { Box } from "@shared/components/Box";
import { BodySmallB } from "@shared/components/Typography";
import { useMemo } from "react";

type SmallSystemIconProps = React.ComponentProps<typeof SmallSystemIcon>;

export type PrefixProps = {
  /* Optional prefix icon to display i.e. AlertIcon */
  prefixIcon?: SmallSystemIconProps["icon"];
  /* Optional prefix icon fill color */
  prefixIconFill?: SmallSystemIconProps["fill"];
  /* Optional prefix text to display i.e. Â£ */
  prefixText?: string;
};

const PrefixIcon: React.FC<SmallSystemIconProps> = (props) => (
  <SmallSystemIcon {...props} w={20} h={20} />
);

export const Prefix: React.FC<PrefixProps> = ({
  prefixIcon,
  prefixIconFill,
  prefixText,
}) => {
  const conditionalContent = useMemo(() => {
    if (prefixText) return <BodySmallB>{prefixText}</BodySmallB>;
    if (prefixIcon)
      return <PrefixIcon icon={prefixIcon} fill={prefixIconFill} />;
    return null;
  }, [prefixIcon, prefixIconFill, prefixText]);

  // Add additional padding when no prefix content
  if (!conditionalContent) return null;

  return (
    <Box
      d="flex"
      align="center"
      px={16}
      py={13}
      backgroundColor="grey6"
      mr={-8}
    >
      {conditionalContent}
    </Box>
  );
};
