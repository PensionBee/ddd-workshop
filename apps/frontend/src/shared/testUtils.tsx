import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
