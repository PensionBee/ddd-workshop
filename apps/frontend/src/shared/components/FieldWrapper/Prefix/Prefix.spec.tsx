import { render } from "@shared/testUtils";
import { vi } from "vitest";
import { Prefix } from "./Prefix";

const PREFIX_TEXT = "Prefix text";
const PREFIX_ICON = "AlertInfo";

describe("FieldWrapper/Prefix", () => {
  describe("Text", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <Prefix prefixText={PREFIX_TEXT} />
      );

      expect(getByText(PREFIX_TEXT)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
    it("should not render if no prefix", () => {
      const { container } = render(<Prefix />);

      expect(container.firstChild).toBeNull();
    });
  });

  describe("Icon", () => {
    it("should render correctly", () => {
      const { container, asFragment } = render(
        <Prefix prefixIcon={PREFIX_ICON} />
      );
      const svgContainer = container.querySelector(
        `[data-rc="SmallSystemIcon.${PREFIX_ICON}"]`
      );

      expect(svgContainer).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
    it("should not render if no prefix", () => {
      const { container } = render(<Prefix />);

      expect(container.firstChild).toBeNull();
    });
  });
});
