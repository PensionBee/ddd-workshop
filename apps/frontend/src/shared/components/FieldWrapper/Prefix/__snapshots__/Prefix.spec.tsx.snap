// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FieldWrapper/Prefix > Text > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnScm fSENfT"
  >
    <span
      class="sc-dkzCjb cGViLk"
      data-rc="BodySmallB"
    >
      Prefix text
    </span>
  </div>
</DocumentFragment>
`;

exports[`FieldWrapper/Prefix Icon should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR iKcDWZ"
  >
    <span
      class="sc-bcXHqe lnJndv"
      data-rc="SmallSystemIcon.AlertInfo"
    >
      <svg>
        alert-info.svg
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`FieldWrapper/Prefix Text should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR iKcDWZ"
  >
    <span
      class="sc-dkrFOg lcwtgS"
      data-rc="BodySmallB"
    >
      Prefix text
    </span>
  </div>
</DocumentFragment>
`;
